File: src/spice_load.rs
use std::collections::BTreeMap;
use std::fs::File;
use std::io::{BufRead, BufReader};

const SKIP: f64 = 2.0;

#[derive(Default)]
pub struct SpiceRead {
    quantum: f64,
    sample_point: f64,
    stability_check: bool,

    num_vars: usize,
    num_points: usize,
    vars: BTreeMap<String, usize>,
    pub index: Vec<usize>,
    raw_values: Vec<f64>,
}

impl SpiceRead {
    pub fn new(quantum: f64, sample_point: f64, stability_check: bool)
               -> SpiceRead {
        SpiceRead {
            quantum, sample_point, stability_check, .. SpiceRead::default()
        }
    }

    pub fn from_args(quantum: f64) -> SpiceRead {
        let args: Vec<String> = std::env::args().collect();
        let mut opts = getopts::Options::new();
        opts.optopt("V", "", "Spice file to verify", "PATH");
        opts.optopt("t", "", "", "");               // Ignore for compatibility.
        let matches = match opts.parse(&args[1..]) {
            Ok(m) => m,
            Err(f) => panic!("{}", f),
        };
        let mut r = SpiceRead::new(quantum, quantum * 0.7, true);
        if let Some(path) = matches.opt_str("V") {
            r.spice_read(&mut BufReader::new(File::open(path).unwrap()))
        }
        else {
            r.spice_read(&mut BufReader::new(std::io::stdin()));
        }
        r
    }

    pub fn spice_read(&mut self, f: &mut impl BufRead) {
        let mut line = String::new();
        loop {
            getline(&mut line, f);

            if line.starts_with("Title: ") || line.starts_with("Date: ")
                || line.starts_with("Plotname:") || line.starts_with("Flags:") {
            }
            else if let Some(num) = line.strip_prefix("No. Variables: ") {
                self.num_vars = num.trim().parse().expect("Bogus num vars");
            }
            else if let Some(num) = line.strip_prefix("No. Points: ") {
                self.num_points = num.trim().parse().expect("Bogus num points");
            }
            else if line.starts_with("Variables:") {
                self.read_var_list(f)
            }
            else if line.starts_with("Binary:") {
                break;
            }
            else {
                panic!("Unexpected line {}", line);
            }
        }

        let total = self.num_points * self.num_vars;
        self.raw_values.resize(total, 0.0);
        {
            let mut slice = unsafe {
                std::slice::from_raw_parts_mut(
                    self.raw_values.as_mut_ptr() as *mut u8,
                    total * std::mem::size_of::<f64>())
            };
            f.read_exact(&mut slice).unwrap();
        }
    
        // Validate the timestamps.
        let mut last = 0.0;
        for t in self.iterate_column("time") {
            assert!(t - last <= 0.21 * self.quantum,
                    "Timestep too large {} {}", last, t);
            assert!(t > last, "Time goes backwards {} {}", last, t);
            last = t;
        }

        // Select the indices to use for sampling.
        let mut last_i = -1;
        let mut index = Vec::new();
        for (row, t) in self.iterate_column("time").enumerate() {
            let i = ((t - self.sample_point) / self.quantum).floor() as isize;
            if i <= last_i {
                continue;
            }
            assert_eq!(i, last_i + 1);
            assert_eq!(i as usize, index.len());
            last_i = i;
            index.push(row * self.num_vars);
        }
        self.index = index;
    }

    pub fn read_var_list(&mut self, f: &mut impl BufRead) {
        let mut line = String::new();
        for i in 0..self.num_vars {
            getline(&mut line, f);
            let mut parts = line.trim_end().split('\t');
            let p1 = parts.next();
            if !p1.is_some_and(|x| x == "") {
                panic!("Corrupt var line '{:?}'", p1);
            }
            let index: usize = parts.next().expect("Truncated var line").parse()
                .expect("Corrupt index");
            assert_eq!(index, i);
            let mut name  = parts.next().expect("Truncated var line");
            let _type = parts.next().expect("Truncated var line");
            parts.next().ok_or(()).expect_err("Trailing junk");
            if let Some(n) = name.strip_prefix("v(") {
                if let Some(n) = n.strip_suffix(")") {
                    name = n;
                }
            }
            self.vars.insert(name.into(), index);
        }
    }

    pub fn iterate_column(&self, name: &str)
                          -> impl '_ + Iterator<Item=f64> {
        let column = self.vars[name];
        self.raw_values[column..].iter().step_by(self.num_vars).copied()
    }

    pub fn extract_sample(&self, name: &str) -> Vec<f64> {
        let column = self.vars[name];
        self.index.iter().map(move |i| self.raw_values[i + column]).collect()
    }

    pub fn extract_signal(&self, name: &str) -> Vec<bool> {
        let column = self.vars[name];
        if self.stability_check {
            self.stability_check(column);
        }
        self.index.iter().map(move |i| self.raw_values[i + column] > 1.2)
            .collect()
    }

    pub fn extract_invert(&self, name: &str) -> Vec<bool> {
        self.extract_signal(name).iter().map(|b| !b).collect()
    }

    pub fn extract_byte(&self, name: &str) -> Vec<u8> {
        let names: [String; 8] = std::array::from_fn(
            |i| format!("{}{}", name, i));
        let columns: [_; 8] = std::array::from_fn(|i| self.vars[&names[i]]);
        let mut res = Vec::new();
        for i in &self.index {
            res.push(columns.iter().enumerate()
                     .map(|(c,b)| (self.raw_values[i+c] > 2.0) as (u8) << b)
                     .sum());
        }
        res
    }

    fn stability_check(&self, column: usize) {
        let tc = self.vars["time"];
        for row in self.raw_values.chunks_exact(self.num_vars) {
            let q = row[tc] / self.quantum;
            let part = q.fract();
            if part < 0.5 || part > 0.9 || q < SKIP * self.quantum {
                continue;
            }
            let val = row[column];
            assert!(val < 0.3 || val > 2.0);
        }
    }
}

fn getline(line: &mut String, f: &mut impl BufRead) {
    line.clear();
    if f.read_line(line).unwrap() == 0 {
        panic!("Unexpected EOF");
    }
}

File: lib/spice_load.cc
#include "spice_load.h"

#include <assert.h>
#include <err.h>
#include <getopt.h>
#include <math.h>
#include <sstream>
#include <string.h>

std::vector<double> spice_load::extract_raw_column(const char * name)
{
    auto p = variables.find(name);
    if (p == variables.end())
        errx(1, "Did not find '%s'", name);
    int col = p->second;
    std::vector<double> r(num_points);
    for (int i = 0; i != num_points; ++i)
        r[i] = raw_values[i * num_vars + col];
    return r;
}


std::vector<bool> spice_load::extract_signal(const char * name)
{
    const auto raw = extract_raw_column(name);
    std::vector<bool> sig(num_samples);
    for (int i = 0; i != num_samples; ++i)
        sig[i] = raw[indexes[i]] > 1.2;
    // Now check that the signal levels are stable between 0.5 QUANTUM and 0.9
    // QUANTUM...
    if (!STABILITY_CHECK)
        return sig;

    for (int i = 0; i != num_points; ++i) {
        int q = timestamps[i] / QUANTUM;
        double remainder = timestamps[i] - q * QUANTUM;
        if (remainder < 0.5 || remainder > 0.9 || q < SKIP * QUANTUM)
            continue;

        if (sig[q] && raw[i] < 3)
            errx(1, "At quantum %i, high signal droop %e", q + 1, raw[i]);
        if (!sig[q] && raw[i] > 0.3)
            errx(1, "At quantum %i, low signal bounce %e", q + 1, raw[i]);
    }
    return sig;
}


std::vector<uint8_t> spice_load::extract_byte(const char * p, const char * s)
{
    std::vector<uint8_t> r(num_samples);
    for (int i = 0; i != 8; ++i) {
        int bit = 1 << i;
        std::ostringstream name;
        name << p << i << s;
        const auto v = extract_signal(name.str().c_str());
        for (int j = 0; j < num_samples; ++j)
            if (v[j])
                r[j] |= bit;
    }
    return r;
}


void spice_load::read_var_list(FILE * f)
{
    char * line = NULL;
    size_t linesize = 0;
    for (long i = 0; i < num_vars; ++i) {
        ssize_t len = getline(&line, &linesize, f);
        if (len < 0)
            errx(1, "Failed to read variables");
        int num;
        char * name;
        char * type;
        if (sscanf(line, "%u%ms%ms", &num, &name, &type) != 3)
            errx(1, "Dodgy variable line: %s", line);
        if (num != i)
            errx(1, "Out of sequence variable line: %s", line);
        // Rewrite v(x) to x.
        int l = strlen(name);
        std::string n = name;
        if (name[0] == 'v' && name[1] == '(' && name[l-1] == ')')
            n = std::string(name + 2, l - 3);
        variables.emplace(n, num);
        free(name);
        free(type);
    }
    free(line);
}


spice_load::spice_load(FILE * f, double q, double sp, bool st) :
    QUANTUM(q),
    SAMPLE_POINT(sp),
    STABILITY_CHECK(st),
    num_vars(0),
    num_points(0)
{
    char * line = NULL;
    size_t linesize = 0;
    while (true) {
        size_t len = getline(&line, &linesize, f);
        if (len <= 0)
            errx(1, "Unexpected eof");
        if (!len || strlen(line) != len || line[len-1] != '\n'
            || strchr(line, ':') == NULL)
            errx(1, "Dodgy line: %s", line);
        if (strncmp(line, "Title: ", 7) == 0)
            ;
        else if (strncmp(line, "Date: ", 6) == 0)
            ;
        else if (strncmp(line, "No. Variables: ", 15) == 0)
            num_vars = strtoul(line + 15, NULL, 0);
        else if (strncmp(line, "No. Points: ", 12) == 0)
            num_points = strtoul(line + 12, NULL, 0);
        else if (strncmp(line, "Variables:", 10) == 0)
            read_var_list(f);
        else if (strncmp(line, "Binary:", 7) == 0)
            break;
    }
    free(line);

    raw_values.resize(num_vars * num_points);
    if (fread(&raw_values[0], num_vars * num_points * sizeof(double),
              1, f) != 1)
        errx(1, "Failed to read data block");

    timestamps = extract_raw_column("time");

    // Validate the timestamps.
    double last = 0;
    for (int i = 0; i != num_points; ++i) {
        if (timestamps[i] - last > 0.21 * QUANTUM)
            errx(1, "Timestep #%i is too large: %e %e\n",
                 i, last, timestamps[i]);
        if (timestamps[i] < last)
            errx(1, "Time goes backwards at %i: %e %e\n",
                 i, last, timestamps[i]);
        last = timestamps[i];
    }

    // Select the indexes to use for sampling.
    int last_index = -1;
    for (int i = 0; i != num_points; ++i) {
        int index = floor((timestamps[i] - SAMPLE_POINT) / QUANTUM);
        if (index < 0 || index == last_index)
            continue;
        assert(index == last_index + 1);
        assert(index == (int) indexes.size());
        last_index = index;
        indexes.push_back(i);
    }
    num_samples = indexes.size();
}


FILE * spice_load::parse_args(int argc, char * const argv[])
{
    const char * path = NULL;
    int c;
    while ((c = getopt(argc, argv, "V:t:")) > 0) {
        switch (c) {
        case 'V':
            path = optarg;
            break;
        case 't':
            break;                      // Ignore for compatibility.
        default:
            errx(1, "Usage: %s [-V path]\n", argv[0]);
            break;
        }
    }
    if (path == NULL)
        return stdin;

    FILE * f = fopen(path, "r");
    if (f == NULL)
        err(1, "opening %s failed", path);
    return f;
}
